integrations/pgvector/src/haystack_integrations/document_stores/pgvector/document_store.py

```diff
@@ -0,0 +1,380 @@
+        )
+
+        if actual_hnsw_index_creation_kwargs:
+            actual_hnsw_index_creation_kwargs_str = ", ".join(
+                f"{key} = {value}" for key, value in actual_hnsw_index_creation_kwargs.items()
+            )
+            sql_add_creation_kwargs = SQL("WITH ({creation_kwargs_str})").format(
+                creation_kwargs_str=SQL(actual_hnsw_index_creation_kwargs_str)
+            )
+            sql_create_index += sql_add_creation_kwargs
+
+        self._execute_sql(sql_create_index, error_msg="Could not create HNSW index")
+
+    def count_documents(self) -> int:
+        """
+        Returns how many documents are present in the document store.
+        """
+
+        sql_count = SQL("SELECT COUNT(*) FROM {table_name}").format(table_name=Identifier(self.table_name))
+
+        count = self._execute_sql(sql_count, error_msg="Could not count documents in PgvectorDocumentStore").fetchone()[
+            0
+        ]
+        return count
+
+    def filter_documents(self, filters: Optional[Dict[str, Any]] = None) -> List[Document]:  # noqa: ARG002
+        # TODO: implement filters
+        sql_get_docs = SQL("SELECT * FROM {table_name}").format(table_name=Identifier(self.table_name))
+
+        result = self._execute_sql(
+            sql_get_docs, error_msg="Could not filter documents from PgvectorDocumentStore", cursor=self._dict_cursor
+        )
+
+        # Fetch all the records
+        records = result.fetchall()
+        docs = self._from_pg_to_haystack_documents(records)
+        return docs
+
+    def write_documents(self, documents: List[Document], policy: DuplicatePolicy = DuplicatePolicy.NONE) -> int:
+        """
+        Writes documents into to PgvectorDocumentStore.
+
+        :param documents: A list of Documents to write to the document store.
+        :param policy: The duplicate policy to use when writing documents.
+        :raises DuplicateDocumentError: If a document with the same id already exists in the document store
+             and the policy is set to DuplicatePolicy.FAIL (or not specified).
+        :return: The number of documents written to the document store.
+        """
+
+        if len(documents) > 0:
+            if not isinstance(documents[0], Document):
+                msg = "param 'documents' must contain a list of objects of type Document"
+                raise ValueError(msg)
+
+        if policy == DuplicatePolicy.NONE:
+            policy = DuplicatePolicy.FAIL
+
+        db_documents = self._from_haystack_to_pg_documents(documents)
+
+        columns_str = "(" + ", ".join(col for col in COLUMNS) + ")"
+        values_placeholder_str = "VALUES (" + ", ".join(f"%({col})s" for col in COLUMNS) + ")"
+
+        insert_statement = SQL("INSERT INTO {table_name} " + columns_str + " " + values_placeholder_str).format(
+            table_name=Identifier(self.table_name)
+        )
```

Comment:
This is very hard to read, I would explicit the query in a string variable `INSERT_STATEMENT` if possible